### PR TITLE
[RFC] fi_verbs: Implement fi_getname() and fi_getpeer()

### PR DESCRIPTION
**Pull request by patrickmacarthur**
I am sending this PR since UNH EXS requires this functionality.

There are two issues with this PR that may require some changes (maybe @shefty has some input on these):
1. The librdmacm calls do not allow the caller to discover the length of the address buffer that it returns, so this code guesses based on `sa_family`.
2. I'm not sure what to do in the case that the address is not available (e.g., since the endpoint isn't connected yet).  FWIW, I don't think librdmacm does anything in this case either, so this code basically just passes whatever librdmacm does up to the user.

### CODE REVIEW

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -1618,6 +1618,71 @@ static struct fi_ops_atomic fi_ibv_msg_ep_atomic_ops = {
...
+{
+	size_t src_addrlen;
+
+	src_addrlen = fi_ibv_getaddrlen(src_addr);
+	if (*dst_addrlen < src_addrlen) {
+		memcpy(dst_addr, src_addr, *dst_addrlen);
+		*dst_addrlen = src_addrlen;
+	} else {
+		memcpy(dst_addr, src_addr, src_addrlen);
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: I think you need to return the correct size of the address in either case.  I.e. move the *dst_addrlen =   portion outside of the if -else clause

File: prov/verbs/src/fi_verbs.c [hunk 2]
```diff
@@ -1618,6 +1618,71 @@ static struct fi_ops_atomic fi_ibv_msg_ep_atomic_ops = {
...
+	return 0;
+}
+
+static int fi_ibv_msg_ep_getname(fid_t ep, void *addr, size_t *addrlen)
+{
+	struct fi_ibv_msg_ep *_ep;
+	struct sockaddr *sa;
+
+	if (*addrlen == 0) {
+		*addrlen = sizeof(struct sockaddr_in);
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Call fi_ibv_getaddrlen() to get the correct size here.

File: prov/verbs/src/fi_verbs.c [hunk 3]
```diff
@@ -1618,6 +1618,71 @@ static struct fi_ops_atomic fi_ibv_msg_ep_atomic_ops = {
...
+
+static int fi_ibv_msg_ep_getpeer(struct fid_ep *ep, void *addr, size_t *addrlen)
+{
+	struct fi_ibv_msg_ep *_ep;
+	struct sockaddr *sa;
+
+	if (*addrlen == 0) {
+		*addrlen = sizeof(struct sockaddr_in);
+		return -FI_ETOOSMALL;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: same as previous comment

File: prov/verbs/src/fi_verbs.c [hunk 4]
```diff
@@ -2576,6 +2641,21 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
...
 
+static int fi_ibv_pep_getname(fid_t pep, void *addr, size_t *addrlen)
+{
+	struct fi_ibv_pep *_pep;
+	struct sockaddr *sa;
+
+	if (*addrlen == 0) {
+		*addrlen = sizeof(struct sockaddr_in);
+		return -FI_ETOOSMALL;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: and here

**Comment by patrickmacarthur** [comment]
The pull request has been updated to take into account @shefty's comments.  In doing so, I have removed a bit more duplication, including the duplication of the pre-existing `fi_ibv_sockaddr_len` function.

**Comment by shefty** [comment]
thanks

**Pull request merged**